In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
from sklearn.datasets import fetch_california_housing
data=fetch_california_housing()
df=pd.DataFrame(data.data,columns=data.feature_names)
df


MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0      8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1      8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2      7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3      5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4      3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   
...       ...       ...       ...        ...         ...       ...       ...   
20635  1.5603      25.0  5.045455   1.133333       845.0  2.560606     39.48   
20636  2.5568      18.0  6.114035   1.315789       356.0  3.122807     39.49   
20637  1.7000      17.0  5.205543   1.120092      1007.0  2.325635     39.43   
20638  1.8672      18.0  5.329513   1.171920       741.0  2.123209     39.43   
20639  2.3886      16.0  5.254717   1.162264      1387.0  2.616981     39.37   

       Longitude  
0        -122.23  
1        -122.22  
2        -122.24  
3        -122.25  
4        -122.25  
...          ...  
20635    -121.09  
20636    -121.21  
20637    -121.22  
20638    -121.32  
20639    -121.24  

[20640 rows x 8 columns]

In [ ]:
df['Target']=data.target
df

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0      8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1      8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2      7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3      5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4      3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   
...       ...       ...       ...        ...         ...       ...       ...   
20635  1.5603      25.0  5.045455   1.133333       845.0  2.560606     39.48   
20636  2.5568      18.0  6.114035   1.315789       356.0  3.122807     39.49   
20637  1.7000      17.0  5.205543   1.120092      1007.0  2.325635     39.43   
20638  1.8672      18.0  5.329513   1.171920       741.0  2.123209     39.43   
20639  2.3886      16.0  5.254717   1.162264      1387.0  2.616981     39.37   

       Longitude  Target  
0        -122.23   4.526  
1        -122.22   3.585  
2        -122.24   3.521  
3        -122.25   3.413  
4        -122.25   3.422  
...          ...     ...  
20635    -121.09   0.781  
20636    -121.21   0.771  
20637    -121.22   0.923  
20638    -121.32   0.847  
20639    -121.24   0.894  

[20640 rows x 9 columns]

In [ ]:
df.columns

Index(['MedInc', 'HouseAge', 'AveRooms', 'AveBedrms', 'Population', 'AveOccup',
       'Latitude', 'Longitude', 'Target'],
      dtype='object')

In [ ]:
df.describe()

MedInc      HouseAge      AveRooms     AveBedrms    Population  \
count  20640.000000  20640.000000  20640.000000  20640.000000  20640.000000   
mean       3.870671     28.639486      5.429000      1.096675   1425.476744   
std        1.899822     12.585558      2.474173      0.473911   1132.462122   
min        0.499900      1.000000      0.846154      0.333333      3.000000   
25%        2.563400     18.000000      4.440716      1.006079    787.000000   
50%        3.534800     29.000000      5.229129      1.048780   1166.000000   
75%        4.743250     37.000000      6.052381      1.099526   1725.000000   
max       15.000100     52.000000    141.909091     34.066667  35682.000000   

           AveOccup      Latitude     Longitude        Target  
count  20640.000000  20640.000000  20640.000000  20640.000000  
mean       3.070655     35.631861   -119.569704      2.068558  
std       10.386050      2.135952      2.003532      1.153956  
min        0.692308     32.540000   -124.350000      0.149990  
25%        2.429741     33.930000   -121.800000      1.196000  
50%        2.818116     34.260000   -118.490000      1.797000  
75%        3.282261     37.710000   -118.010000      2.647250  
max     1243.333333     41.950000   -114.310000      5.000010

In [ ]:
df.isna().sum()

MedInc        0
HouseAge      0
AveRooms      0
AveBedrms     0
Population    0
AveOccup      0
Latitude      0
Longitude     0
Target        0
dtype: int64



#Outlier Handling


In [ ]:
q1,q2,q3=np.percentile(df['MedInc'],[25,50,75])
iqr=q3-q1
lower_bound=q1-(1.5*iqr)
upper_bound=q3+(1.5*iqr)
print(lower_bound,upper_bound)
outlier=[]
for i in df['MedInc']:
  if i<lower_bound or i>upper_bound:
    outlier.append(i)
print(outlier)
df['MedInc']=np.where(df['MedInc']>upper_bound,upper_bound,df['MedInc'])
df['MedInc']=np.where(df['MedInc']<lower_bound,lower_bound,df['MedInc'])


-0.7063750000000004 8.013024999999999
[8.3252, 8.3014, 11.6017, 8.2049, 8.401, 8.317, 8.1564, 8.8793, 8.8342, 10.0825, 9.3959, 11.8603, 13.499, 12.2138, 8.1872, 12.3804, 8.1194, 8.7477, 8.3996, 8.4045, 9.7194, 8.3337, 10.9506, 10.3203, 9.1531, 9.5862, 8.155, 8.3188, 9.7037, 10.3345, 12.5915, 15.0001, 9.8708, 8.2673, 8.4352, 10.7372, 13.4883, 9.5271, 12.2478, 8.0839, 9.0776, 8.039, 10.4549, 10.3224, 8.3678, 8.0448, 10.3713, 11.7064, 11.3421, 8.1586, 8.9571, 8.7589, 10.9405, 10.3416, 9.0064, 8.1489, 10.5815, 13.2949, 13.1499, 8.3209, 8.3956, 8.1163, 9.0243, 8.7059, 10.5144, 8.9669, 8.7716, 8.3608, 8.2737, 10.0263, 10.5941, 8.5325, 9.2873, 8.1248, 8.8467, 8.1125, 10.8805, 8.2159, 8.176, 10.9052, 8.203, 8.1356, 9.6047, 8.565, 8.1714, 8.7471, 8.8612, 8.6454, 8.3841, 11.2093, 8.4438, 9.6465, 8.3912, 10.9201, 8.4196, 10.5511, 8.7172, 13.4196, 10.3587, 10.6796, 11.2979, 12.8879, 9.2189, 10.2292, 10.4399, 10.0549, 8.2269, 10.3599, 8.8115, 8.153, 10.9237, 8.2954, 9.042, 8.0499, 8.5677, 10.5575, 

In [ ]:
q1,q2,q3=np.percentile(df['HouseAge'],[25,50,75])
iqr=q3-q1
lower_bound=q1-(1.5*iqr)
upper_bound=q3+(1.5*iqr)
print(lower_bound,upper_bound)
outlier=[]
for i in df['HouseAge']:
  if i<lower_bound or i>upper_bound:
    outlier.append(i)
print(outlier)
df['HouseAge']=np.where(df['HouseAge']>upper_bound,upper_bound,df['HouseAge'])
df['HouseAge']=np.where(df['HouseAge']<lower_bound,lower_bound,df['HouseAge'])


-10.5 65.5
[]


In [ ]:
q1,q2,q3=np.percentile(df['AveRooms'],[25,50,75])
iqr=q3-q1
lower_bound=q1-(1.5*iqr)
upper_bound=q3+(1.5*iqr)
print(lower_bound,upper_bound)
outlier=[]
for i in df['AveRooms']:
  if i<lower_bound or i>upper_bound:
    outlier.append(i)
print(outlier)
df['AveRooms']=np.where(df['AveRooms']>upper_bound,upper_bound,df['AveRooms'])
df['AveRooms']=np.where(df['AveRooms']<lower_bound,lower_bound,df['AveRooms'])

2.023219161170969 8.469878027106942
[1.7142857142857142, 8.972868217054264, 8.928358208955224, 9.210227272727273, 9.122715404699738, 9.0, 8.518248175182482, 8.973684210526315, 10.51530612244898, 29.852941176470587, 8.62530413625304, 10.168591224018476, 31.77777777777778, 16.847533632286996, 28.75728155339806, 13.473039215686274, 11.490861618798956, 9.793028322440087, 28.112903225806452, 33.98969072164948, 47.515151515151516, 8.527293844367016, 9.263598326359833, 8.481045751633987, 8.606299212598426, 8.94675925925926, 8.74031007751938, 10.511627906976743, 22.22222222222222, 9.016112789526687, 9.361702127659575, 8.540925266903914, 9.104216388225934, 8.802531645569621, 8.853061224489796, 23.88888888888889, 9.77578475336323, 38.84090909090909, 13.104166666666666, 10.639455782312925, 8.886039886039885, 20.900783289817234, 11.183266932270916, 9.308383233532934, 11.679878048780488, 9.979166666666666, 9.418685121107266, 11.024561403508772, 10.375478927203066, 9.053571428571429, 16.883561643835

In [ ]:
q1,q2,q3=np.percentile(df['AveBedrms'],[25,50,75])
iqr=q3-q1
lower_bound=q1-(1.5*iqr)
upper_bound=q3+(1.5*iqr)
print(lower_bound,upper_bound)
outlier=[]
for i in df['AveBedrms']:
  if i<lower_bound or i>upper_bound:
    outlier.append(i)
print(outlier)
df['AveBedrms']=np.where(df['AveBedrms']>upper_bound,upper_bound,df['AveBedrms'])
df['AveBedrms']=np.where(df['AveBedrms']<lower_bound,lower_bound,df['AveBedrms'])

0.8659085155701288 1.2396965968190603
[]


In [ ]:
q1,q2,q3=np.percentile(df['Population'],[25,50,75])
iqr=q3-q1
lower_bound=q1-(1.5*iqr)
upper_bound=q3+(1.5*iqr)
print(lower_bound,upper_bound)
outlier=[]
for i in df['Population']:
  if i<lower_bound or i>upper_bound:
    outlier.append(i)
print(outlier)
df['Population']=np.where(df['Population']>upper_bound,upper_bound,df['Population'])
df['Population']=np.where(df['Population']<lower_bound,lower_bound,df['Population'])

-620.0 3132.0
[3469.0, 4367.0, 4985.0, 3337.0, 3276.0, 3668.0, 3741.0, 7427.0, 5781.0, 3827.0, 3827.0, 5613.0, 4065.0, 5436.0, 3388.0, 3174.0, 3163.0, 8117.0, 8012.0, 3436.0, 5019.0, 4576.0, 6075.0, 7205.0, 4446.0, 3288.0, 4518.0, 3386.0, 4649.0, 4552.0, 3215.0, 12203.0, 3571.0, 4319.0, 3526.0, 3223.0, 3386.0, 3329.0, 3926.0, 4847.0, 4389.0, 5826.0, 3200.0, 8768.0, 3489.0, 6693.0, 4930.0, 4644.0, 3628.0, 4030.0, 4591.0, 7517.0, 3316.0, 3264.0, 5022.0, 3175.0, 3266.0, 3578.0, 4460.0, 3523.0, 4276.0, 3838.0, 3207.0, 3266.0, 3221.0, 4377.0, 3748.0, 3621.0, 4741.0, 3645.0, 3264.0, 3486.0, 4089.0, 5613.0, 3221.0, 3283.0, 3280.0, 4083.0, 3371.0, 3325.0, 7653.0, 3898.0, 3596.0, 3326.0, 7266.0, 3637.0, 3259.0, 3408.0, 3546.0, 5001.0, 3952.0, 3530.0, 5034.0, 4179.0, 3786.0, 4749.0, 3573.0, 5561.0, 4476.0, 3406.0, 3407.0, 3415.0, 5466.0, 4638.0, 5226.0, 4432.0, 3535.0, 3565.0, 3242.0, 4781.0, 4584.0, 7009.0, 4250.0, 5696.0, 4870.0, 3278.0, 3525.0, 6241.0, 4611.0, 4101.0, 4893.0, 3269.0, 4163.0, 

In [ ]:
q1,q2,q3=np.percentile(df['AveOccup'],[25,50,75])
iqr=q3-q1
lower_bound=q1-(1.5*iqr)
upper_bound=q3+(1.5*iqr)
print(lower_bound,upper_bound)
outlier=[]
for i in df['AveOccup']:
  if i<lower_bound or i>upper_bound:
    outlier.append(i)
print(outlier)
df['AveOccup']=np.where(df['AveOccup']>upper_bound,upper_bound,df['AveOccup'])
df['AveOccup']=np.where(df['AveOccup']<lower_bound,lower_bound,df['AveOccup'])

1.1509614824735064 4.5610405893536905
[4.658823529411765, 10.272727272727273, 5.617647058823529, 4.666666666666667, 12.23404255319149, 5.467153284671533, 5.1753554502369665, 5.396, 9.534285714285714, 5.738028169014084, 17.177700348432055, 11.295918367346939, 4.8370044052863435, 12.212765957446809, 6.076923076923077, 6.061538461538461, 4.825, 4.942084942084942, 4.773809523809524, 4.608695652173913, 4.743473325766175, 5.520833333333333, 4.663865546218488, 5.206422018348624, 4.693009118541034, 4.809523809523809, 5.072664359861592, 5.769230769230769, 5.702247191011236, 4.897260273972603, 4.653950953678474, 5.2504672897196265, 4.866666666666666, 5.03030303030303, 5.286307053941909, 5.2, 14.0, 6.0, 12.84375, 4.907801418439717, 5.052132701421801, 4.75, 4.725589225589226, 4.993265993265993, 4.67816091954023, 6.076923076923077, 11.634146341463415, 5.2635983263598325, 6.497584541062802, 4.728110599078341, 4.860126582278481, 5.2622377622377625, 4.621621621621622, 9.4, 599.7142857142857, 4.5894465

In [ ]:
q1,q2,q3=np.percentile(df['Latitude'],[25,50,75])
iqr=q3-q1
lower_bound=q1-(1.5*iqr)
upper_bound=q3+(1.5*iqr)
print(lower_bound,upper_bound)
outlier=[]
for i in df['Latitude']:
  if i<lower_bound or i>upper_bound:
    outlier.append(i)
print(outlier)
df['Latitude']=np.where(df['Latitude']>upper_bound,upper_bound,df['Latitude'])
df['Latitude']=np.where(df['Latitude']<lower_bound,lower_bound,df['Latitude'])

28.259999999999998 43.38
[]


In [ ]:
q1,q2,q3=np.percentile(df['Longitude'],[25,50,75])
iqr=q3-q1
lower_bound=q1-(1.5*iqr)
upper_bound=q3+(1.5*iqr)
print(lower_bound,upper_bound)
outlier=[]
for i in df['Longitude']:
  if i<lower_bound or i>upper_bound:
    outlier.append(i)
print(outlier)
df['Longitude']=np.where(df['Longitude']>upper_bound,upper_bound,df['Longitude'])
df['Longitude']=np.where(df['Longitude']<lower_bound,lower_bound,df['Longitude'])

-127.48499999999999 -112.32500000000002
[]


In [ ]:
q1,q2,q3=np.percentile(df['Target'],[25,50,75])
iqr=q3-q1
lower_bound=q1-(1.5*iqr)
upper_bound=q3+(1.5*iqr)
print(lower_bound,upper_bound)
outlier=[]
for i in df['Target']:
  if i<lower_bound or i>upper_bound:
    outlier.append(i)
print(outlier)
df['Target']=np.where(df['Target']>upper_bound,upper_bound,df['Target'])
df['Target']=np.where(df['Target']<lower_bound,lower_bound,df['Target'])

-0.9808749999999995 4.824124999999999
[5.00001, 4.833, 5.00001, 4.896, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 4.94, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 4.99, 5.00001, 5.00001, 5.00001, 4.854, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 4.835, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 4.831, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 4.955, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 4.908, 5.00001, 5.00001, 4.956, 5.00001, 5.00001, 4.838, 5.00001, 4.976, 5.00001, 5.00001, 4.956, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 4.987, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5.00001, 5

In [ ]:
df.skew()

MedInc        0.735618
HouseAge      0.060331
AveRooms      0.348567
AveBedrms     0.462645
Population    0.842247
AveOccup      0.510453
Latitude      0.465953
Longitude    -0.297801
Target        0.912330
dtype: float64

In [ ]:
X=df.drop('Target',axis=1)
y=df['Target']

In [ ]:
from sklearn.preprocessing import StandardScaler
scaler=StandardScaler()
X=scaler.fit_transform(X)
X=pd.DataFrame(X,columns=data.feature_names)


In [ ]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.3,random_state=42)


#Linear Regression


In [ ]:
from sklearn.linear_model import LinearRegression
lr=LinearRegression()
lr.fit(X_train,y_train)

LinearRegression()

In [ ]:
y_pred=lr.predict(X_test)
y_pred

array([0.30860733, 1.79493817, 3.18467386, ..., 2.35732213, 1.39269536,
       2.0086847 ])

In [ ]:
y_test

20046    0.477000
3024     0.458000
15663    4.824125
20484    2.186000
9814     2.780000
           ...   
17505    2.375000
13512    0.673000
10842    2.184000
16559    1.194000
5786     2.098000
Name: Target, Length: 6192, dtype: float64

In [ ]:
slope = lr.coef_
intercept = lr.intercept_

print("Slope of best fit line is:", slope)
print("Intercept of best fit line is:", intercept)

slope_df = pd.DataFrame(slope.reshape(1, -1), columns=data.feature_names).T
slope_df.columns = ['Slope']
print(slope_df)

equation = "y = " + str(intercept)
for feature, coef in zip(data.feature_names, slope):
    equation += f" + ({coef})*{feature}"

print("Equation of best fit line:")
print(equation)


Slope of best fit line is: [ 0.76024602  0.15263063 -0.08290086  0.09979408  0.05445711 -0.28205724
 -0.87295525 -0.79703403]
Intercept of best fit line is: 2.059403961633797
               Slope
MedInc      0.760246
HouseAge    0.152631
AveRooms   -0.082901
AveBedrms   0.099794
Population  0.054457
AveOccup   -0.282057
Latitude   -0.872955
Longitude  -0.797034
Equation of best fit line:
y = 2.059403961633797 + (0.7602460238154438)*MedInc + (0.15263062879905837)*HouseAge + (-0.08290086498513084)*AveRooms + (0.0997940754602408)*AveBedrms + (0.05445711227903434)*Population + (-0.28205723849549147)*AveOccup + (-0.8729552492174765)*Latitude + (-0.7970340310698873)*Longitude


Linear Regression works by fitting a straight line that best describes the relationship between the input features and the target variable, minimizing the difference between predicted and actual values. It is suitable for this dataset because it provides a simple, interpretable baseline model and works well when the relationships between variables are mostly linear. Although not the most accurate, it helps establish a foundation for comparing more complex algorithms.

In [ ]:
from sklearn.metrics import mean_squared_error,r2_score,mean_absolute_error
print(mean_squared_error(y_test,y_pred))
print(r2_score(y_test,y_pred))
print(mean_absolute_error(y_test,y_pred))

0.4411480790926837
0.6514877387669372
0.4962950068093782


A multiple linear regression model was fitted on the dataset to predict the target variable using all available features. The model achieved an R² score of 0.65, indicating that approximately 65% of the variance in the target variable is explained by the model.

The error metrics show that the model performs reasonably well:

Mean Squared Error (MSE): 0.44

Mean Absolute Error (MAE): 0.49

These values indicate that the model’s predictions are fairly close to the actual values on average, though there is still room for improvement. Overall, the regression model provides a good baseline performance and offers meaningful insights into the relationship between the input features and the target variable.

#KNN

In [ ]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.3,random_state=42)

In [ ]:
from sklearn.neighbors import KNeighborsRegressor
knn=KNeighborsRegressor(p=2,n_neighbors=9)
knn.fit(X_train,y_train)

KNeighborsRegressor(n_neighbors=9)

In [ ]:
knn=KNeighborsRegressor(p=2,n_neighbors=5)
knn.fit(X_train,y_train)

KNeighborsRegressor()

K-Nearest Neighbors (KNN) predicts a value by looking at the ‘k’ closest data points in the feature space and averaging their target values. It works well for this dataset because it can capture nonlinear relationships without assuming any mathematical form. KNN is especially useful when similar data points tend to have similar outputs, as seen in many real-world datasets.

In [ ]:
y_new=knn.predict(X_test)
y_new



array([0.50577778, 1.14466667, 4.80666667, ..., 2.20244444, 1.39322222,
       2.05322222])

In [ ]:
y_k5=knn.predict(X_test)
y_k5

array([0.5066  , 1.258   , 4.824125, ..., 2.3906  , 1.2244  , 1.9444  ])

In [ ]:
print(r2_score(y_test,y_new))
print(mean_squared_error(y_test,y_new))
print(mean_absolute_error(y_test,y_new))

0.7368195019816628
0.33313482499774655
0.40076437930663217


In [ ]:
print(r2_score(y_test,y_k5))
print(mean_squared_error(y_test,y_k5))
print(mean_absolute_error(y_test,y_k5))

0.7211740293596298
0.35293892075405475
0.40767153682170537


A K-Nearest Neighbors (KNN) regression model was evaluated using different values of k to determine the most suitable neighborhood size for prediction.

For k = 9, the model achieved:

R² Score: 0.7368

Mean Squared Error (MSE): 0.3331

Mean Absolute Error (MAE): 0.4008

For k = 5, the performance slightly decreased:

R² Score: 0.7212

MSE: 0.3529

MAE: 0.4077

These results indicate that k = 9 provides better predictive performance compared to k = 5, with a higher R² score and lower error values. Overall, increasing the neighborhood size to 9 improves the model's stability and accuracy for this dataset.



#Decision Tree Regressor

In [ ]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.3,random_state=42)


In [ ]:
from sklearn.tree import DecisionTreeRegressor
dt=DecisionTreeRegressor(random_state=1)
dt.fit(X_train,y_train)

DecisionTreeRegressor(random_state=1)

Decision Tree Regression works by splitting the data into smaller subsets based on feature values, forming a tree-like structure where each split aims to reduce prediction error. It is suitable for this dataset because it can capture nonlinear patterns and interactions between features without requiring any scaling. However, it may overfit, which is why its performance can be less stable compared to ensemble methods.

In [ ]:
y_pred_dt=dt.predict(X_test)
y_pred_dt

array([0.425   , 0.702   , 4.824125, ..., 2.13    , 1.293   , 2.583   ])

In [ ]:
print(r2_score(y_test,y_pred_dt))
print(mean_squared_error(y_test,y_pred_dt))
print(mean_absolute_error(y_test,y_pred_dt))

0.5866732926649904
0.5231904390778583
0.4707264228036175


A Decision Tree Regressor was trained and evaluated on the dataset. The model achieved an R² score of 0.5867, indicating that it explains about 59% of the variance in the target variable.

The error metrics are as follows:

Mean Squared Error (MSE): 0.5232

Mean Absolute Error (MAE): 0.4707

Compared to other models such as Linear Regression and KNN, the Decision Tree shows lower predictive accuracy and higher error values, suggesting that the tree may be overfitting the training data or failing to generalize well to unseen samples. While it captures nonlinear relationships, its overall performance is less stable and less accurate for this dataset.

##Random forest Regressor

In [ ]:
from sklearn.ensemble import RandomForestRegressor
rf=RandomForestRegressor(random_state=1)
rf.fit(X_train,y_train)

RandomForestRegressor(random_state=1)

Random Forest Regression builds multiple decision trees on different subsets of the data and averages their predictions to improve accuracy and reduce overfitting. It is suitable for this dataset because it captures complex, nonlinear relationships and handles noise effectively. Its ensemble nature makes it more stable and reliable than a single decision tree.

In [ ]:
y_pred_rf=rf.predict(X_test)
y_pred_rf

array([0.49277 , 0.73554 , 4.712885, ..., 2.07636 , 1.46454 , 2.23906 ])

In [ ]:
print(r2_score(y_test,y_pred_rf))
print(mean_squared_error(y_test,y_pred_rf))
print(mean_absolute_error(y_test,y_pred_rf))

0.8080429094527741
0.24297997856235953
0.3261006601178941


A Random Forest Regressor was applied to the dataset to improve prediction accuracy by combining multiple decision trees. The model performed significantly better than the previous models, achieving an R² score of 0.8080, which means it explains about 81% of the variance in the target variable.

The error metrics further show strong performance:

Mean Squared Error (MSE): 0.2430

Mean Absolute Error (MAE): 0.3261

These values indicate that the Random Forest model provides more accurate and stable predictions compared to Linear Regression, KNN, and Decision Tree models. Its ensemble nature helps reduce overfitting and captures complex relationships in the data, making it the best-performing model among those evaluated.

#Support Vector Machine

In [ ]:
from sklearn.svm import SVR
svr=SVR()
svr.fit(X_train,y_train)

SVR()

Support Vector Regression (SVR) works by finding a curve or line that best fits the data while keeping errors within a small margin, using kernels to model complex nonlinear relationships. It is suitable for this dataset because it handles nonlinear patterns well and is robust to outliers. SVR often performs strongly when the data has a mix of linear and nonlinear trends.

In [ ]:
y_pred_svr=svr.predict(X_test)
y_pred_svr

array([0.50204206, 1.45395867, 4.66181367, ..., 2.08885391, 1.27943294,
       1.93027284])

In [ ]:
print(r2_score(y_test,y_pred_svr))
print(mean_squared_error(y_test,y_pred_svr))
print(mean_absolute_error(y_test,y_pred_svr))

0.7712395301982023
0.2895658289563474
0.36481132347002854


A Support Vector Regression (SVR) model was trained to capture nonlinear relationships in the dataset. The model achieved a strong R² score of 0.7712, meaning it explains approximately 77% of the variance in the target variable.

The error metrics are as follows:

Mean Squared Error (MSE): 0.2896

Mean Absolute Error (MAE): 0.3648

These results indicate that the SVR model performs well, offering better accuracy than Decision Tree and Linear Regression models. However, its performance is slightly lower than that of the Random Forest model, which remains the best among the tested algorithms. Overall, SVR provides a good balance between generalization and capturing complex patterns in the data.

# Gradient Booster

In [ ]:
from sklearn.ensemble import GradientBoostingRegressor
gb=GradientBoostingRegressor()
gb.fit(X_train,y_train)

GradientBoostingRegressor()

Gradient Boosting Regression builds a series of weak decision trees, where each new tree learns from the errors of the previous ones to gradually improve prediction accuracy. It is suitable for this dataset because it captures complex nonlinear patterns and reduces bias effectively. This sequential learning approach often leads to strong, high-accuracy performance on tabular data.

In [ ]:
y_pred_gb=gb.predict(X_test)
y_pred_gb

array([0.57558382, 1.03750612, 4.01722415, ..., 2.10969191, 1.36982176,
       1.92646328])

In [ ]:
print(r2_score(y_test,y_pred_gb))
print(mean_squared_error(y_test,y_pred_gb))
print(mean_absolute_error(y_test,y_pred_gb))

0.7832013664740585
0.27442449339229896
0.3651293926453479


A Gradient Boosting Regressor was trained to improve prediction accuracy by combining multiple weak learners in a sequential manner. The model achieved a strong R² score of 0.7832, indicating that it explains about 78% of the variance in the target variable.

The error metrics are as follows:

Mean Squared Error (MSE): 0.2744

Mean Absolute Error (MAE): 0.3651

These results show that Gradient Boosting performs better than Linear Regression, Decision Tree, and KNN models, and slightly better than SVR. While its performance is strong, it still does not surpass the Random Forest Regressor, which remains the best-performing model in this project. Overall, Gradient Boosting provides accurate predictions and captures complex relationships effectively

#Conclusion
Several regression models were evaluated to identify the most effective algorithm for predicting the target variable. Among all models, the Random Forest Regressor continues to be the best performer, achieving the highest R² score (0.808) and the lowest error values (MSE: 0.243, MAE: 0.326). Its ensemble approach reduces overfitting and captures complex interactions, making it the most reliable model.

The Gradient Boosting Regressor ranks second, demonstrating strong predictive performance with an R² score of 0.783 and low error metrics. It effectively captures nonlinear patterns by boosting weak learners sequentially.

SVR and KNN (k=9) offer good performance, while Linear Regression provides a baseline for comparison. The Decision Tree model shows the weakest generalization ability due to overfitting.

Overall, the project shows that ensemble models—especially Random Forest and Gradient Boosting—are the most effective choices for this regression task, offering the highest accuracy and stability.